src/diff/mod.rs

```diff
@@ -1,36 +1,269 @@
+                },
+                (None, Some(new_entry)) => {
+                    diff.add_created_files(new_entry, parent_path)?;
+                    new_entry_opt = new_iter.next();
+                },
+                (None, None) => break
+            }
+        }
+        Ok(())
+    }
+
+    // if entry is a file, then return this file,
+    // or a list of files in the directory tree otherwise
+    fn collect_files_from_entry<F, T>(entry: &DirectoryContents, parent_path: &Rc<RefCell<Vec<Label>>>, mapper: F)
+        -> Result<Vec<T>, String>
+        where F: Fn(&File, &Vec<Label>) -> T + Copy {
+
+        match entry {
+            DirectoryContents::SubDirectory(dir) => Diff::collect_files(dir, parent_path, mapper),
+            DirectoryContents::File(file) => {
+                let mapped = mapper(file, &RefCell::borrow(parent_path));
+                Ok(vec![mapped])
+            },
+            _ => Err(String::from("Unexpected entry type"))
+        }
+    }
+
+    fn collect_files<F, T>(dir: &Directory, parent_path: &Rc<RefCell<Vec<Label>>>, mapper: F)
+        -> Result<Vec<T>, String>
+        where F: Fn(&File, &Vec<Label>) -> T + Copy {
+
+        let mut files: Vec<T> = Vec::new();
+        Diff::collect_files_inner(dir, parent_path, mapper, &mut files)?;
+        Ok(files)
+    }
+
+    fn collect_files_inner<'a, F, T>(dir: &'a Directory, parent_path: &Rc<RefCell<Vec<Label>>>, mapper: F, files: &mut Vec<T>)
+        -> Result<(), String>
+        where F: Fn(&File, &Vec<Label>) -> T + Copy {
+
+        parent_path.borrow_mut().push(dir.label.clone());
+        for entry in dir.entries.iter() {
+            match entry {
+                DirectoryContents::SubDirectory(subdir) => {
+                    Diff::collect_files_inner(subdir.deref(), parent_path, mapper, files)?;
+                },
+                DirectoryContents::File(file) => {
+                    files.push(mapper(file, &RefCell::borrow(parent_path)));
+                },
+                DirectoryContents::Repo => { /* Skip repo directory */ }
+            }
+        }
+        parent_path.borrow_mut().pop();
+        Ok(())
+    }
+
+    fn convert_to_deleted(file: &File, parent_path: &Vec<Label>) -> DeleteFile {
+        DeleteFile {
+            path: Path::from_labels(file.filename.clone(), &parent_path)
+        }
+    }
+
+    fn convert_to_created(file: &File, parent_path: &Vec<Label>) -> CreateFile {
+        CreateFile {
+            path: Path::from_labels(file.filename.clone(), parent_path)
+        }
+    }
+
+    fn add_modified_file(&mut self, file: &File, parent_path: &Vec<Label>) {
+        // TODO: file diff can be calculated at this point
+        // Use pijul's transaction diff as an inspiration?
+        // https://nest.pijul.com/pijul_org/pijul:master/1468b7281a6f3785e9#anesp4Qdq3V
+        self.modified.push(ModifiedFile {
+            path: Path::from_labels(file.filename.clone(), parent_path),
```

Comment:
I think you want `parent_path.push(file.filename.clone())`, that's why you're getting the reversal

Reply:
I suggest to discuss the issue of modifying Path in the comment above

Reply:
:+1: